### PR TITLE
Make ReplayStage create the parallel fork replay threadpool

### PR DESCRIPTION
**Pull request by steviez**
#### Problem
As part of work for https://github.com/anza-xyz/agave/issues/105, it will be desirable to make the thread pool size configurable on the CLI. The pool is currently within a `lazy_static`; configuring the size for that from the CLI is a little "messy". We'd need to add another function that sets a global (before the thread pool is created), and then have thread pool creation read that value. This is of course susceptible to a potential race condition and/or some annoying code to ensure that the pool comes back the size we actually wanted it to be.

#### Summary of Changes
Instead of using the lazy_static, just have `ReplayStage` create the thread pool (if parallel replay is enabled).

There will be some whitespace difference - I'd suggest either viewing with "hide whitespace", or view commit-by-commit (I ran `cargo fmt` in separate commits)

### CODE REVIEW

File: core/src/replay_stage.rs [hunk 1]
```diff
@@ -3186,11 +3193,17 @@ impl ReplayStage {
             num_active_banks,
             active_bank_slots
         );
-        if num_active_banks > 0 {
-            let replay_result_vec = if num_active_banks > 1 && replay_slots_concurrently {
+        if active_bank_slots.is_empty() {
+            return false;
+        }
```
Comment (steviez) [inline core/src/replay_stage.rs:3198]: I thought that this was a simple enough sanity check to pull out and explicitly return early for - this enabled us to remove a level of indentation for everything else

Comment (bw-solana) [reply core/src/replay_stage.rs:3198]: Never nesters unite

**Review by bw-solana — APPROVED** [review]
LGTM!

**Comment by bw-solana** [comment]
How painful is it on validators to switch up CLI args? It would be nice to replace `replay_slots_concurrently` with `replay_forks_concurrently` to better convey what we're actually doing

**Comment by steviez** [comment]
> How painful is it on validators to switch up CLI args? It would be nice to replace `replay_slots_concurrently` with `replay_forks_concurrently` to better convey what we're actually doing

I was probably going to partially deprecate this argument anyways in favor an argument that allows for configuration of the number of threads. So, something like:
```
--concurrent-fork-replay-threads N
```
We can still accept the old flag (but issue a warning that it is deprecated), but just choose a reasonable default value for the old flag (since the old flag is a bool, not a number)

File: core/src/replay_stage.rs (hunk 1 above)
Comment (t-nelson) [reply core/src/replay_stage.rs:3198]: early returns make code more complex and easier to introduce bugs in the future. CHANGE MY MIND

Comment (bw-solana) [reply core/src/replay_stage.rs:3198]: Number of bugs in a function is proportional to the maximum indent. CHANGE MY MIND

Comment (diman-io) [reply core/src/replay_stage.rs:3198]: 🍿

Comment (steviez) [reply core/src/replay_stage.rs:3198]: > early returns make code more complex and easier to introduce bugs in the future

https://github.com/steviez/solana/commit/4bcbbbe825f38f998f54f22329a75c9d7ddfc2ce?diff=unified&w=0

Comment (t-nelson) [reply core/src/replay_stage.rs:3198]: > Number of bugs in a function is proportional to the maximum indent. CHANGE MY MIND

you're stuck in the present old man. join us in the future!

**Pull request merged**